src/adapters/bridge-api.ts

```diff
@@ -28,139 +27,205 @@ interface MerkleProof {
+      rollup_exit_root: z.string().length(66),
+    })
+    .transform(apiMerkleProofToDomain)
 );
 
-const getMerkleProofResponseParser = StrictSchema<GetMerkleProofResponse>()(
+const getMerkleProofResponseParser = StrictSchema<
+  {
+    proof: MerkleProof;
+  },
+  {
+    proof: domain.MerkleProof;
+  }
+>()(
   z.object({
     proof: merkleProofParser,
   })
 );
 
-const claimParser = StrictSchema<Claim>()(
-  z.object({
-    index: z.string(),
-    token_addr: z.string(),
-    amount: z.string(),
-    dest_net: z.number(),
-    dest_addr: z.string(),
-    block_num: z.string(),
-  })
+const apiClaimToDomain = ({ index, block_num }: Claim): domain.Claim => ({
+  index: z.number().nonnegative().parse(Number(index)),
+  blockNumber: block_num,
+});
+
+const claimParser = StrictSchema<Claim, domain.Claim>()(
+  z
+    .object({
+      index: z.string(),
+      block_num: z.string(),
+    })
+    .transform(apiClaimToDomain)
 );
 
-const getClaimsResponseParser = StrictSchema<GetClaimsResponse>()(
+const getClaimsResponseParser = StrictSchema<
+  {
+    claims?: Claim[];
+  },
+  {
+    claims?: domain.Claim[];
+  }
+>()(
   z.object({
     claims: z.optional(z.array(claimParser)),
   })
 );
 
-const apiBridgeToDomain = ({
-  token_addr,
-  amount,
-  dest_addr,
-  dest_net,
-  deposit_cnt,
-}: Bridge): domain.Bridge => ({
-  tokenAddress: token_addr,
-  amount: amount,
-  destinationAddress: dest_addr,
-  destinationNetwork: dest_net,
-  depositCount: deposit_cnt,
-});
+interface GetTransactionsParams {
+  env: domain.Env;
+  ethereumAddress: string;
+}
 
-const apiMerkleProofToDomain = ({
-  merkle_proof,
-  exit_root_num,
-  main_exit_root,
-  rollup_exit_root,
-}: MerkleProof): domain.MerkleProof => ({
-  merkleProof: merkle_proof,
-  exitRootNumber: exit_root_num,
-  mainExitRoot: main_exit_root,
-  rollupExitRoot: rollup_exit_root,
-});
+const getTransactions = async ({
+  env,
+  ethereumAddress,
+}: GetTransactionsParams): Promise<domain.Transaction[]> => {
+  const apiUrl = env.bridgeApiUrl;
+  const [bridges, claims] = await Promise.all([
+    getBridges({ apiUrl, ethereumAddress }),
+    getClaims({ apiUrl, ethereumAddress }),
+  ]);
 
-const apiClaimToDomain = ({
-  index,
-  token_addr,
-  amount,
-  dest_net,
-  dest_addr,
-  block_num,
-}: Claim): domain.Claim => ({
-  index,
-  tokenAddress: token_addr,
-  amount,
-  destinationNetwork: dest_net,
-  destinationAddress: dest_addr,
-  blockNumber: block_num,
-});
+  return await Promise.all(
+    bridges.map(async (bridge): Promise<domain.Transaction> => {
+      const {
+        originNetwork,
+        destinationNetwork,
+        networkId,
+        amount,
+        destinationAddress,
+        depositCount,
+      } = bridge;
+      const [claimStatus, merkleProof]: [domain.ClaimStatus, domain.MerkleProof] =
+        await Promise.all([
+          getClaimStatus({ apiUrl, networkId, depositCount }),
+          getMerkleProof({ apiUrl, networkId, depositCount }),
```

Comment:
We shouldn't get the `MerkleProof` until the user does a `claim`.

Doing this, we will request the merkle proof even if the `getClaimStatus` returns `false`. I would request the proof only when the user wants to click on the `Finalise` button to avoid unnecessary calls to the API.

At least, I would request it only when the `getClaimStatus` returns `true`, but this doesn't guarantee that the user is going to use the proof at all (as he only needs it for claiming).

Reply:
Yep, this makes a lot of sense. To be honest I focused on making the thing work and didn't even think about optimizing the flow of requests.

I've removed the `MerkleProof` from the already claimed TXs (`status: "completed"`) since once a TX has been claimed we no longer need to get the `MerkleProof`. This alone will prevent obtaining the `MerkleProof` for all the historic TXs.

Also in `getTransactions` now the first thing we do is check whether the TX has been claimed. If a TX has been already claimed we can return right away without even having to call `claimStatus`.

If the TX can't be found in the claim list, then we call `getClaimStatus` and return right away when the claim is not yet ready.

And finally, only when the claim is ready we do call `getMerkleProof` and return a TX with `status: "on-hold"`.

Given that the list of TXs pending to be claimed is expected to be small and that the `MerkleProof` is a lightweight payload, I think this level of optimization is more than enough (i.e. I'd not obtain the `MerkleProof` when the user clicks the Claim button)